src/validators/shapes.rs

```diff
@@ -17,7 +19,54 @@ pub fn validate(gtfs: &gtfs_structures::Gtfs) -> Vec<Issue> {
             Issue::new(Severity::Error, IssueType::InvalidCoordinates, id)
                 .object_type(gtfs_structures::ObjectType::Shape)
         });
-    missing_coord.chain(valid).collect()
+    let invalid_shape_id = gtfs
+        .trips
+        .iter()
+        .map(|(_id, trip)| create_invalid_shape_id_issue(trip, gtfs))
+        .filter_map(|o| o);
+
+    let used_shape_id: HashSet<String> = gtfs
+        .trips
+        .iter()
+        .map(|(_trip_id, trip)| trip.shape_id.to_owned())
+        .filter_map(|o| o)
+        .collect();
+
+    let existing_shape_id: HashSet<String> = gtfs
```

Comment:
not very important, but you could remove a layer of string allocation by doing:

```rust

    let used_shape_id: HashSet<&str> = gtfs
        .trips
        .iter()
        .filter_map(|(_trip_id, trip)| trip.shape_id.as_ref().map(|s| s.as_str()))
        .collect();

    let existing_shape_id: HashSet<&str> = gtfs
        .shapes
        .iter()
        .map(|(shape_id, _shape)| shape_id.as_str())
        .collect();
```

Reply:
Thanks! I didn't even need to use `as_str()`, and it seems to work.